File: examples/java/cart/src/main/java/dev/angzarr/examples/cart/CartState.java
package dev.angzarr.examples.cart;

import java.util.HashMap;
import java.util.Map;

public record CartState(
    String customerId,
    Map<String, CartItem> items,
    int subtotalCents,
    String couponCode,
    int discountCents,
    String status
) {
    public static CartState empty() {
        return new CartState("", new HashMap<>(), 0, "", 0, "");
    }

    public boolean exists() {
        return !customerId.isEmpty();
    }

    public boolean isCheckedOut() {
        return "checked_out".equals(status);
    }

    public int calculateSubtotal() {
        return items.values().stream()
            .mapToInt(item -> item.quantity() * item.unitPriceCents())
            .sum();
    }

    public record CartItem(String productId, String name, int quantity, int unitPriceCents) {}
}

File: examples/kotlin/cart/src/main/kotlin/dev/angzarr/examples/cart/CartState.kt
package dev.angzarr.examples.cart

import examples.Domains.LineItem

data class CartState(
    val customerId: String = "",
    val items: List<LineItem> = emptyList(),
    val subtotalCents: Int = 0,
    val couponCode: String = "",
    val discountCents: Int = 0,
    val status: String = ""
) {
    fun exists(): Boolean = customerId.isNotEmpty()
    fun isActive(): Boolean = status == "active"
    fun isCheckedOut(): Boolean = status == "checked_out"

    companion object {
        fun empty(): CartState = CartState()
    }
}
